utils/cli/multisend.go

```diff
@@ -0,0 +1,419 @@
+					txs, err = authtx.QueryTxsByEvents(clientCtxOrigin, tmEvents, query.DefaultPage, query.DefaultLimit, "")
+					if err == nil {
+						break
+					}
+				}
+				if err != nil {
+					fmt.Printf("failed to query tx %s\n", err)
+					return false
+				}
+				if len(txs.Txs) == 0 {
+					fmt.Println("found no txs matching given address and sequence combination")
+					return false
+				}
+				if len(txs.Txs) > 1 {
+					// This case means there's a bug somewhere else in the code. Should not happen.
+					fmt.Printf("found %d txs matching given address and sequence combination\n", len(txs.Txs))
+					return false
+				}
+
+				return txs.Txs[0].Code == 0
+			}
+
+			waitSequenceChange := func(account string, sequence uint64) error {
+				currentSequence, err := getSequence(account)
+				if err != nil {
+					return err
+				}
+				for currentSequence < sequence {
+					fmt.Printf("waiting for sequence number %d current %d \n", sequence, currentSequence)
+					time.Sleep(5 * time.Second)
+
+					currentSequence, err = getSequence(account)
+					if err != nil {
+						return err
+					}
+				}
+				if !getResponse(account, sequence-1) {
+					return fmt.Errorf("transaction failed")
+				}
+				return nil
+			}
+
+			if progress.SequenceOrigin == 0 {
+				progress.SequenceOrigin, err = getSequence(clientCtxOrigin.FromAddress.String())
+				if err != nil {
+					return err
+				}
+			}
+
+			if progress.SequenceHotWallet == 0 {
+				progress.SequenceHotWallet = progress.SequenceOrigin
+				if useHotWallet {
+					progress.SequenceHotWallet, err = getSequence(clientCtxHotWallet.FromAddress.String())
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			output := []banktypes.Output{}
+			totalAmount := sdk.Coins{}
+			records = records[progress.Index:]
+			for i, record := range records {
+				coins, err := sdk.ParseCoinsNormalized(record[1])
+				if err != nil {
+					fmt.Printf("failed sending records from %d to %d\n", progress.Index, i)
+					fmt.Printf("please run again with: multi-send [file.csv] %d\n", progress.Index)
+					return err
+				}
+
+				if coins.IsZero() {
+					fmt.Printf("failed sending records from %d to %d\n", progress.Index, i)
+					fmt.Printf("please run again with: multi-send [file.csv] %d\n", progress.Index)
+					return fmt.Errorf("must send positive amount")
+				}
+				toAddr, err := sdk.AccAddressFromBech32(record[0])
+				if err != nil {
+					fmt.Printf("failed sending records from %d to %d\n", progress.Index, i)
+					fmt.Printf("please run again with: multi-send [file.csv] %d\n", progress.Index)
+					return err
+				}
+				output = append(output, banktypes.NewOutput(toAddr, coins))
+				totalAmount = totalAmount.Add(coins...)
+
+				if (i+1)%MAX_ADDRESSES == 0 || (i+1) == len(records) {
+					if useHotWallet {
+						if progress.Progress == PRG_ready {
+							msg := banktypes.NewMsgSend(clientCtxOrigin.FromAddress, clientCtxHotWallet.FromAddress, totalAmount.Add(fees...))
```

Comment:
does not check sequence before sending, saved progress and on chain sequence might be different